### ISSUE TITLE
Java.Net.ProtocolException when using AndroidClientHandler: PROPFIND (and others) unsupported

### ISSUE THREAD

**Issue by tipa** [skazantsev/WebDavClient#59]
I am using this library in my multi-platform Xamarin app and attempt to use the native HttpClientHandler on each platform (Windows -> HttpClientHandler, iOS -> NSUrlSessionHandler, Android -> AndroidClientHandler).

In the case of Android:
```
var httpHandler = new AndroidClientHandler() { AutomaticDecompression = DecompressionMethods.Deflate | DecompressionMethods.GZip };
httpHandler.Credentials = new NetworkCredential(credentials.Username, credentials.Password);
var client = new WebDavClient(new HttpClient(httpHandler, true) { BaseAddress = new Uri(credentials.URL) });
```

Unfortunately, this causes the following exception when using the `WebDavClient`:
`Java.Net.ProtocolException: Expected one of [OPTIONS, GET, HEAD, POST, PUT, DELETE, TRACE, PATCH] but was PROPFIND`

A few possible solutions/workarounds are mentioned here: https://stackoverflow.com/questions/25163131/httpurlconnection-invalid-http-method-patch (X-HTTP-Method-Override, Reflection, ...)

I understand that this is not a bug in your library, but maybe you are interested in adding a workaround anyways...

**Comment by pp111** [comment]
I am in the same situation, the AndroidClientHandler is not working with PROPFIND and the HttpClientHandler is not working with some certificates. So it is not possible to use this library to access some servers.

**Comment by tipa** [comment]
> So it is not possible to use this library to access some servers.

@pp111 it is possible, but you need to use `HttpClientHandler` instead of `AndroidClientHandler `

**Comment by pp111** [comment]
If I use HttpClientHandler I get the exception 

Authentication failed, see inner exception. ---> Mono.Btls.MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED 

The certificate is fine, but needs TLS 1.2 which seems to be not supported with HttpClientHandler on Android (on Windows it works fine).

Using AndroidClientHandler the certificate validation works fine, but it fails because of the PROPFIND issue.

So, if I am not missing something, none of the two handlers are working with TLS1.2 servers on Android.

**Comment by tipa** [comment]
You can ignore that certificate check like this:
```
var httpHandler = new HttpClientHandler()
{
    Credentials = new NetworkCredential(usernam, password),
    ServerCertificateCustomValidationCallback = (httpRequestMessage, cert, cetChain, policyErrors) => true,
};
```

**Comment by pp111** [comment]
Yes, I think this is the only option although it might be potentially unsafe.

**Comment by tipa** [comment]
The `HttpClientHandler` stopped working with .net 6 Android. It is now required to opt-out of the `AndroidMessageHandler` on a project-level. I reported this issue the Microsoft here: https://github.com/xamarin/xamarin-android/issues/7291

**Comment by cesarchefinho** [comment]
> The `HttpClientHandler` stopped working with .net 6 Android. It is now required to opt-out of the `AndroidMessageHandler` on a project-level. I reported this issue the Microsoft here: [xamarin/xamarin-android#7291](https://github.com/xamarin/xamarin-android/issues/7291)

cN you post here an step by step how-to make work with android ? I try all things and cannot resolve the proppath exception...

thanks

**Comment by tipa** [comment]
@cesarchefinho 

```
var httpHandler = new SocketsHttpHandler();
httpHandler.AutomaticDecompression = DecompressionMethods.Deflate | DecompressionMethods.GZip;
httpHandler.Credentials = new NetworkCredential(username, password);
var client = new WebDavClient(new HttpClient(httpHandler, true) { BaseAddress = new Uri(url) });
```

**Comment by tipa** [comment]
> After some consideration I decided to not include a workaround into the library since it's a rare case and it should be possible to implement a workaround on the client (hopefully `X-HTTP-Method-Override` is enough).

I am still forced to use a `SocketsHttpHandler` in order to send `PROPFIND` requests on Android, but this comes at other disadvantages (exceptions, performance). `AndroidMessageHandler` is just the recommended http handler to be used on Android. Can you expand on your comment how one can use `X-HTTP-Method-Override` with the current implementation to overcome this problem?

**Comment by igiona** [comment]
> var httpHandler = new SocketsHttpHandler();

@tipa does the `SocketsHttpHandler` stil work on net8 (android) for you?
For me every request times out, whereas `HttpClientHandler` works at least for file uploads.

**Comment by tipa** [comment]
Yes, SocketsHttpHandler generally still works for me

**Comment by igiona** [comment]
Interesting,

I'm connecting to a next-cloud instance, with a app-password scheme so I need to set the auth via headers as follow

``` 
var encodedCredentials = Convert.ToBase64String(Encoding.UTF8.GetBytes(userName + ":" + appPassword));
httpClient.DefaultRequestHeaders.Add("OCS-APIRequest", "true");
httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", encodedCredentials);
```

...pherhaps is somehow related to this difference.
The annoying thing is that the request just stalls, doesn't return "Unauthorized" or so.

**Comment by igiona** [comment]
> Yes, SocketsHttpHandler generally still works for me

@tipa may I ask what kind of Android version have you tested?
`SocketsHttpHandler` doesn't work on my physical devices running Android14 doesn't work as described above, whereas it works fine on an Emulator running Android12.
Admittedly, also the standard handler doesn't work on my physical phone...

**Comment by tipa** [comment]
> > Yes, SocketsHttpHandler generally still works for me
> 
> @tipa may I ask what kind of Android version have you tested?
> `SocketsHttpHandler` doesn't work on my physical devices running Android14 doesn't work as described above, whereas it works fine on an Emulator running Android12.
> Admittedly, also the standard handler doesn't work on my physical phone...

I am using it in production, with basically all Android versions

**Comment by igiona** [comment]
> > > Yes, SocketsHttpHandler generally still works for me
> > 
> > 
> > @tipa may I ask what kind of Android version have you tested?
> > `SocketsHttpHandler` doesn't work on my physical devices running Android14 doesn't work as described above, whereas it works fine on an Emulator running Android12.
> > Admittedly, also the standard handler doesn't work on my physical phone...
> 
> I am using it in production, with basically all Android versions

Okay, thanks.
It must be somewhat something weird with my Phone or something which interferes with it, perhaps some Uno library...
